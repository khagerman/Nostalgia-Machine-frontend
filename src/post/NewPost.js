import React, { useState, useContext } from "react";
import { useHistory } from "react-router-dom";

import NostalgiaApi from "../api";
import LoadingSpinner from "../common/LoadingSpinner";
import UserContext from "../auth/UserContext";
import { Formik, Form, Field } from "formik";
import { object, string } from "yup";
import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import {
  TextField,
  FormGroup,
  Container,
  Button,
  Alert,
  Select,
  MenuItem,
} from "@mui/material";
/**
   *NewPosr component

   *form to create new post
*/

function NewPost() {
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const history = useHistory();

  // seperate state for select input
  const [decade, setDecade] = useState("1");
  let username = currentUser?.username;

  async function handleSubmit(data) {
    try {
      // creates post and updates current user to update users profile
      let result = await NostalgiaApi.postPost(data);
      let currentUser = await NostalgiaApi.getUser(username);
      setCurrentUser(currentUser);
      // goes to decade page of given decade of post
      history.push(`/decade/${parseInt(decade)}`);
      toast("Post created!", {
        position: toast.POSITION.TOP_CENTER,
      });
    } catch (errors) {
      console.error(errors);
      // alerts of errors if any
      toast.error(errors[0], {
        position: toast.POSITION.TOP_CENTER,
      });
    }
  }

  // update seperate state for select input
  const handleChangeSelect = (event) => {
    setDecade(event.target.value);
  };

  //schema for post
  const PostSchema = object({
    url: string().required().url(),
    title: string().required().min(1).max(100),
  });

  return (
    <>
      <h2 className="m-3">New Post/Memory</h2>

      <Formik
        initialValues={{
          url: "",
          title: "",
        }}
        validationSchema={PostSchema}
        onSubmit={(values) => {
          handleSubmit({ ...values, decade_id: parseInt(decade) });
        }}
      >
        {({ errors, touched }) => (
          <Container maxWidth="xs">
            <Form>
              <FormGroup className="mb-2">
                <Field
                  id="url"
                  name="url"
                  className="form-control"
                  label="Image URL"
                  as={TextField}
                />
                {errors.url && touched.url ? (
                  <Alert severity={"error"}>{errors.url}</Alert>
                ) : null}
              </FormGroup>
              <FormGroup className="mb-2">
                <Field
                  name="title"
                  id="title"
                  label="Post Text"
                  className="form-control"
                  as={TextField}
                />

                {errors.title && touched.title ? (
                  <Alert severity={"error"}>{errors.title}</Alert>
                ) : null}
              </FormGroup>
              <FormGroup className="pb-2">
                <Select
                  id="decade"
                  name="decade"
                  value={decade}
                  onChange={handleChangeSelect}
                  label="decade"
                  className="label"
                  style={{ "aria-label": "Without label" }}
                >
                  <MenuItem name="1960s" value="1">
                    1960s
                  </MenuItem>
                  <MenuItem name="1970s" value="2">
                    1970s
                  </MenuItem>
                  <MenuItem name="1980s" value="3">
                    1980s
                  </MenuItem>
                  <MenuItem name="1990s" value="4">
                    1990s
                  </MenuItem>
                  <MenuItem name="2000s" value="5">
                    2000s
                  </MenuItem>
                </Select>
              </FormGroup>

              <Button
                color="secondary"
                variant="contained"
                type="submit"
                onSubmit={handleSubmit}
              >
                Submit
              </Button>
            </Form>
          </Container>
        )}
      </Formik>
    </>
  );
}

export default NewPost;
