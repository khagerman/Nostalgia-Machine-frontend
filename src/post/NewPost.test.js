import React from "react";
import { render } from "@testing-library/react";
import NewPost from "./NewPost";
import { MemoryRouter } from "react-router";
import { UserProvider } from "../testUtils";
it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <NewPost />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
