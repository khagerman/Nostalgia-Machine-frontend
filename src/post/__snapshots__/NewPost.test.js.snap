// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h2
    class="m-3"
  >
    New Post/Memory
  </h2>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <form
      action="#"
    >
      <div
        class="MuiFormGroup-root mb-2 css-dmmspl-MuiFormGroup-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root form-control css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="url"
            id="url-label"
          >
            Image URL
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="url"
              name="url"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Image URL
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiFormGroup-root mb-2 css-dmmspl-MuiFormGroup-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root form-control css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="title"
            id="title-label"
          >
            Post Text
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="title"
              name="title"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Post Text
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiFormGroup-root pb-2 css-dmmspl-MuiFormGroup-root"
      >
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary label css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="decade"
            class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="decade"
            role="button"
            tabindex="0"
          >
            1960s
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="decade"
            tabindex="-1"
            value="1"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-173wfxe"
            >
              <span>
                decade
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-zcbmsk-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</DocumentFragment>
`;
